### PR TITLE
ENH: improve binned_statistics_2d performance.

### LINKED ISSUE #11468: performance of binned_statistics_2d 100x slowdown from 1.3.2 to 1.4.0
My issue is about the performance of scipy.stats.binned_statistic_2d going from 1.3.2 to 1.4.0. 

The function performs hugely worse in calculating the std, so much so that we can't use version 1.4.0. I've seen that the calculation has changed because of #10126, and it looks like for good reason, but is there a way we can use the old calculation?

#### Reproducing code example and timeit results:
```
import scipy.stats
import numpy as np

inp = np.random.rand(9999).reshape(3,3333)*200

subbin_x_edges = np.arange(0, 200, dtype=np.float32)
subbin_y_edges = np.arange(0, 200, dtype=np.float64)

%%timeit
stat='std'
binned_stat, _, _, binnumbers = scipy.stats.binned_statistic_2d(
    inp[0], inp[1], values=inp[2], statistic=stat,
    bins=[subbin_x_edges, subbin_y_edges], expand_binnumbers=True)
```

Scipy 1.3.2
```
1.25 ms ± 5.87 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

Scipy 1.4.0
```
122 ms ± 1.98 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

#### Scipy/Numpy/Python version information:
Scipy 1.4.0 
Numpy 1.18.1 
sys.version_info(major=3, minor=6, micro=9, releaselevel='final', serial=0)

My system: 
Ubuntu 18.04.3 LTS
Memory: 15.5GiB
Processor: Intel® Core™ i7-7700HQ CPU @ 2.80GHz × 8
Graphics: GeForce GTX 1050/PCIe/SSE2
OS type: 64-bit

Have confirmed this issue exists on MacOSX as well.

Thank you for your time
-->

### PR DESCRIPTION
**Pull request by AtsushiSakai**
#### Reference issue
To fix #11468 

#### What does this implement/fix?
Improve binned_statistics_2d performance

### CODE REVIEW

**Comment by AtsushiSakai** [comment]
This is my benchmark code based on #11468 
   ```
 def test_binned_statistic_dd_performance(self):
        import timeit
        def test():
            inp = np.random.rand(9999).reshape(3, 3333) * 200

            subbin_x_edges = np.arange(0, 200, dtype=np.float32)
            subbin_y_edges = np.arange(0, 200, dtype=np.float64)

            stat = 'std'
            binned_stat, _, _, binnumbers = binned_statistic_2d(
                inp[0], inp[1], values=inp[2], statistic=stat,
                bins=[subbin_x_edges, subbin_y_edges], expand_binnumbers=True)
        loop = 100
        result = timeit.timeit(lambda: test(), number=loop)
        print(result / loop)
```

**Comment by AtsushiSakai** [comment]
Improve result:
Before -> 0.14416 sec
After    -> 0.06928 sec

It is improved, but I think it is not enough. 
Does anyone have idea to improve it more?

**Comment by rgommers** [comment]
Thanks @AtsushiSakai. It would be useful to add your benchmark code to `benchmarks/benchmarks/stats.py`, there's no benchmarks for any of the `binned_statistic` functions yet.

@rlucas7 want to review this one?

**Comment by AtsushiSakai** [comment]
@rgommers Thank you for reviewing. I added a benchmark.

**Comment by AtsushiSakai** [comment]
This is the benchmark result of asv:
SS 2020-04-27 at 22 34 14

**Comment by rgommers** [comment]
Hmm, the issue complains about a 100x slowdown, this is a 4x speedup. So this probably isn't enough (just based on the numbers, didn't look at the code).

**Comment by rlucas7** [comment]
I might have time to take a closer look this weekend, probably Sunday.

Sincerely, 

-Lucas Roberts

**Comment by AtsushiSakai** [comment]
> So this probably isn't enough

@rgommers I agree your point, however, as discussed in #11468, the previous version might become numerically unstable. I think this PR is worth until "online" variance calculation is implemented.

**Comment by rlucas7** [comment]
> > So this probably isn't enough
> 
> @rgommers I agree your point, however, as discussed in #11468, the previous version might become numerically unstable. I think this PR is worth until "online" variance calculation is implemented.

To be clear the numerical instability noted in issue #11468 is already fixed in `master`, PR #10351 was the fix, which was originally noted in issue #10126. IN issue #11468 the comment is that the fix slowed down the calculations. I agreed with this assessment of slowdown. 

The tradeoff in PR #10351 was one of in favor of ***correctness*** over speed. The previous implementation was not correct for cases when mean is large relative to variance. As SciPy is a library that aims to be general purpose we cannot really say the `std` calculation only works when variance is large relative to the mean inside `binned_statistic_dd()` routines. 

The only way I know to maintain stability for the `'std'` calculation involve either shifting by a constant to stabilize (usually the sample mean, or to use an 'online' std calculation using a (variation of) Welford's algorithm. 

There are other options like `mpmath` that we wouldn't do inside a publicly exposed SciPy function or assuming the inputs satisfy `mean/variance` is small enough to have a stable condition number for the std calculation. 

The stability is all laid out fairly clearly in [Hingham's book](https://www.amazon.com/Accuracy-Stability-Numerical-Algorithms-Nicholas/dp/0898715210) and also in some other spots on the web including in Knuth's TAOCP (vol 2) which gives the algorithm the name Welford.

File: scipy/stats/_binned_statistic.py
Comment (rlucas7) [inline scipy/stats/_binned_statistic.py:579]: what about the cases when `binned_data < 2` ? currently `numpy.std` returns a `0.0` and this would omit those values entirely and would not be backwards compatible with 1.4.x versions of SciPy.

File: benchmarks/benchmarks/stats.py
Comment (rlucas7) [inline benchmarks/benchmarks/stats.py:164]: This is a useful change to track, thanks for adding the benchmark. 
The `binned_statistic_dd` method supports both a scenario where you pass in data and the bin edges are calculated inside the method as well as the scenario where you've already computed the binedges and would like to re-use them. It would be nice to track both input cases for this method. Tracking both situations will give an indication of the speed up due to changes in std calculation alone versus other changes inside the routine.

**Review by rlucas7 — CHANGES_REQUESTED** [review]
@AtsushiSakai Thanks for the proposed changes. I agree w/Ralf's sentiments on the speedup aspect. The benchmark is fine and useful for us to have in the benchmarks as a number to track. 

However, the current change method changes do not preserve backwards compatibility and omit certain values in the routine's calculations so that part of the PR needs to be removed before we can merge. 

In this case, it might make sense to change the title of the PR from "improve ... performance" to "add benchmarks ..."

File: scipy/stats/_binned_statistic.py
Comment (rlucas7) [reply scipy/stats/_binned_statistic.py:579]: Instead you could add a test case for `std` routine with a binned_data where only 1 entry was in the bin and have that bin calculate a `0.0` value. We don't have one of those currently AFAIK.

**Comment by AtsushiSakai** [comment]
@rlucas7 First of all, thank you so much for taking you time. I'm sorry I was not aware that this PR break backward compatibility. If so, I would like to remove the change.

File: scipy/stats/_binned_statistic.py
Comment (AtsushiSakai) [reply scipy/stats/_binned_statistic.py:579]: In my understanding, all `result` items are initialized with 0.  `numpy.std` returns `0.0` when the number of binned_data is less than 2 (0 or 1). In that cases, we don't need to calculate `np.std`, that makes calculation faster. But my understanding might be wrong...

>  this would omit those values entirely
Could you explain the details of this?

File: benchmarks/benchmarks/stats.py
Comment (AtsushiSakai) [reply benchmarks/benchmarks/stats.py:164]: OK. I will try it.

File: scipy/stats/_binned_statistic.py
Comment (rlucas7) [reply scipy/stats/_binned_statistic.py:579]: >In my understanding, all result items are initialized with 0.

Looking at it again today, you're right. I must've been looking at the `'mean'` case right above it where they default to `nan`s. 

The case when the number of binned data are 0 is perhaps a little less obvious that those would be `0`. I think that should be mentioned explicitly in the docstring. 

I don't think it was possible to do that before we added the functionality to re-use a binned statistic calculation which is a more recent change.

Comment (rlucas7) [reply scipy/stats/_binned_statistic.py:579]: > @rlucas7 First of all, thank you so much for taking your time. I'm sorry I was not aware that this PR break backward compatibility. If so, I would like to remove the change. Before that, please take a look my comments in the reviews.

I left a comment to clarify.

File: benchmarks/benchmarks/stats.py
Comment (AtsushiSakai) [reply benchmarks/benchmarks/stats.py:164]: I tried use `binned_statistic_result` for reused benchmark. But, an error happened and I don't know how to fix it... Could you please help me?

This is my code.
```
import numpy as np
from scipy import stats

np.random.seed(12345678)
inp = np.random.rand(9999).reshape(3, 3333) * 200
subbin_x_edges = np.arange(0, 200, dtype=np.float32)
subbin_y_edges = np.arange(0, 200, dtype=np.float64)
stat = 'std'

ret = stats.binned_statistic_dd(
    [inp[0], inp[1]], inp[2], statistic=stat,
    bins=[subbin_x_edges, subbin_y_edges],
    expand_binnumbers=True)
print(ret)

ret = stats.binned_statistic_dd(
    [inp[0], inp[1]], -inp[2], statistic=stat,
    binned_statistic_result=ret)
print(ret)
```

But, this error happened in the second `binned_statistic_dd` call.

>  File "/home/scipy/build/testenv/lib/python3.6/site-packages/scipy/stats/_binned_statistic.py", line 584, in binned_statistic_dd
>                    flatcount = np.bincount(binnumbers, None)
>                  File "<__array_function__ internals>", line 6, in bincount
>                ValueError: object too deep for desired array

**Comment by rlucas7** [comment]
Difficult to say from looking at it on my Phone but you are using the same value: ret as an argument to second binned statistic and as the returned assignment. Not sure if that is the issue or not but you probably want to change at least. 

The docstring has example usage for reuse of prior result, do you get the same error if you follow the usage in the docstring? 

Sincerely, 

-Lucas Roberts

File: scipy/stats/_binned_statistic.py
Comment (AtsushiSakai) [reply scipy/stats/_binned_statistic.py:580]: > I think that should be mentioned explicitly in the docstring.
I did it.

File: benchmarks/benchmarks/stats.py
Comment (AtsushiSakai) [reply benchmarks/benchmarks/stats.py]: >It would be nice to track both input cases for this method.

I added re-use bin benchmark too.

**Comment by AtsushiSakai** [comment]
I found the reason the error. The reason of this error is `expand_binnumbers=True` in the first `stats.binned_statistic_dd` call.
If using result with `expand_binnumbers=True` for `binned_statistic_result`, the error happens.
I created an issue to improve doc and error message for this (#12115). I will fix it in another PR.

**Comment by AtsushiSakai** [comment]
I updated docstring and added another benchmark code as you requested. PTAL.

File: scipy/stats/_binned_statistic.py
Comment (rlucas7) [inline scipy/stats/_binned_statistic.py:395]:
```suggestion
            is implicitly calculated with ddof=0. If the number of values within a given bin is 0 or 1, 
the computed standard deviation value will be 0 for the bin.
```

if you agree the layout of the suggestion across the lines might need editing to keep within PEP8.

File: benchmarks/benchmarks/stats.py
Comment (rlucas7) [inline benchmarks/benchmarks/stats.py:173]: I'm confused by the naming of the test and the statistic argument here. The test name states test is for std yet the statistic is `"mean"`? What is the reasoning?

**Review by rlucas7 — CHANGES_REQUESTED** [review]
@AtsushiSakai I left a couple comments, getting closer. I'm not clear the logic on one of the tests cases and left a message, can you clarify?

**Comment by rlucas7** [comment]
Not certain but the travis errors seem legitimate but not related to this PR, something to do with pydata changes? 
https://travis-ci.org/github/scipy/scipy/jobs/687008282#L3024

numpy has no gcd attribute, an error in numba:
https://travis-ci.org/github/scipy/scipy/jobs/687008282#L3002
did we bump a numpy version a function got dropped or renamed in numba? 

either that or have we updated changes in the build that would be impacting? (perhaps related to migration?)

File: benchmarks/benchmarks/stats.py
Comment (AtsushiSakai) [reply benchmarks/benchmarks/stats.py]: You are right. I just copy and paste the code from docstring, but we should use `std`. I changed it.

**Comment by tylerjereddy** [comment]
@rlucas7 Are the backwards compatibility concerns addressed now? I see that CI is green--perhaps this one is within reach for branching 1.5.x in ~6 days?

File: scipy/stats/_binned_statistic.py
Comment (rlucas7) [reply scipy/stats/_binned_statistic.py:594]: Yes, Thanks for that. Marking resolved.

**Review by rlucas7 — APPROVED** [review]
Thanks for the changes @AtsushiSakai I'm marking approved. I'll leave up for another couple days and merge this weekend if no other comments are made. 

@tylerjereddy I think this one will be merged by this weekend, would be good to get into the 1.5.0 branch. 

Edit: adding a comment to answer Tyler's question, yes no backwards compatibility concerns that I see.

**Comment by rlucas7** [comment]
@AtsushiSakai squashed and merged. Thanks for the improvements, keep up the good work.

**Pull request merged**
